### ISSUE TITLE
How to initialize cyclical computations?

### ISSUE THREAD

**Issue by fenduru** [adamhaile/S#1]
I'm looking to create a graph of computed data where there may exist dependencies. I had previously begun working on my own library, but came across this one and it seems to handle most of my requirements that other libraries don't, in particular the "no redundant computations" and its ability to handle cycles.

From my experimentation though, the only way to hook up a cycle is if the cycle loops by emitting onto a data signal. This has a few undesirable properties

* A computation needs to know which data signals need to be updated
* Cycles must involve data signals (i.e. can't have interdependent computations)
* Replacing computations with data-signals loses the benefits of S doing dependency tracking (leading to redundant computations, inconsistent states, etc).

Boilerplate of what I'm trying to achieve: https://jsfiddle.net/uto837uf/

Example of a not-working cycle (by creating dependent computations): https://jsfiddle.net/uto837uf/3/

One solution that I can think of to this is to have a way to create a dummy node up front so that you have a reference to use for dependency tracking, and then "replace" the dummy node with the real thing later. I haven't found a way to do this from user-space though.

**Comment by adamhaile** [comment]
Hi @fenduru --
In general, yes, you're right: cyclic computations generally have a data signal somewhere in the path.  In fact, a totally synchronous cycle (one with no data signals in the cycle) is an error, and S will throw an exception saying so.  As you mention, it takes mutating a signal to create one:

```javascript
S.root(() => {
	var a = S.data(1),
  	c = () => 2, // placeholder c()
  	b = S(() => a() + c()); // runs w/ placeholder c()
    
	c = S(() => b() * 2);// real c(), runs b/c b->c dep hasn't been recalculated
  
  a(2); // throws "circular dependency"
});
``` 
S's execution model, what SRP literature calls the "synchronous hypothesis," requires that a signal have exactly one value per "instant."  Cycles violate this: if the *current* value of b() depends on the *current* value of c(), and the current value of c() on the current value of b(), then neither can be determined.

You *can* have cross-dependencies between nodes so long as they aren't both active in the same instant:

```javascript
S.root(() => {
    const a = S.data(false),
    b = S(() => a() ? c() : 1), // doesn't call c() yet, since a() is false, returns 1
    c = S(() => a() ? 2 : b()); // calls b(), returns 2

    a(true); // now c() calls b() but not vice-versa.  both now equal 2
});
```

You can also have cyclic dependencies when the cycle is split across time, aka there's a data signal in the cycle such that the current value of some set of nodes depends on the *last* value of others:

```javascript
S.root(() => {
    const a = S.data(1),
    lastC = S.data(0), // manual starting value
    b = S(() => a() + lastC()),
    c = S(() => b() * 2);
    
    S(() => lastC(c())); // "glue" comp bridges c() and lastC()
});
```

A cycle like the one above, though, has no termination: lastC()is updated each time, so S runs the nodes again ... and again ... and again, etc.  After 100,000 ticks, S will finally throw an exception about a "runaway clock detected."  This is because usually this kind of cycle is a code error.

Temporally split cycles are fine if you put a guard in place though:
```javascript
S.root(() => {
	const a = S.data(1),
  	lastC = S.data(0), // placeholder initial value
    b = S(() => a() + lastC()),
    c = S(() => b() * 2);
    
    S(() => c() < 1000 && lastC(c())); // "guard" prevents infinite updates
});
```
For instance, you could create a cyclic computation that uses the Newtonian method to approximate a value, then exits once the adjustment gets below some threshold.

Does that help?  If there's something you're trying to do with cycles that isn't covered here, tell me more about it and I can see if I can offer any help.

**Comment by fenduru** [comment]
Thanks for the very thorough reply!

> You can also have cyclic dependencies when the cycle is split across time, aka there's a data signal in the cycle such that the current value of some set of nodes depends on the last value of others:

So this is more along the lines of what I'm trying to accomplish, however the problem with this is that since `b` doesn't actually depend on `c` (rather it depends on `lastC`), this causes us to have redundant and/or intermediate computations that get thrown out.

Example:

```js
S.root(() => {
	// `c` is not defined yet because it needs to depend on `b`
	let a = S.data(1);
  let lastC = S.data(0);
  
  let b = S(() => {
  	console.log('calculating b:', a(), '+', lastC());
  	return a() + lastC()
  });
  
  let c = S(() => a() * 2);
  
  S(() => lastC(c()));
  
  console.log('setting a');
  a(2);
});
```

Compare this to the same code that doesn't have the indirection of a data signal:

```js
S.root(() => {
	// `c` is not defined yet because it needs to depend on `b`
	let a = S.data(1);
  let lastC = S.data(0);
  
  let c = S(() => a() * 2);
  
  let b = S(() => {
  	console.log('calculating b:', a(), '+', c());
  	return a() + c()
  });
  
  S(() => lastC(c()));
  
  console.log('setting a');
  a(2);
});
```

After we set `a`, `b` gets computed twice (once with the old C and new A, once with the new C and new A). This is very reminiscent of ["glitches"](https://stackoverflow.com/questions/22332407/how-to-avoid-glitches-in-rx) in libraries like RxJs.

I had been doing some digging into research projects like [timely dataflow](https://github.com/frankmcsherry/timely-dataflow) to gather inspiration, and while the whitepapers are a bit over my head, the general idea of breaking the graph into strong components, and treating each component/iteration of a component as a "subclock" seems promising.

To illustrate, let's say my conceptual graph looks like this:

![image](https://user-images.githubusercontent.com/1354771/27090771-236a2132-502c-11e7-8ca6-ecc01d96e882.png)

We can break this up into strongly connected components, which turns our cyclic graph into a DAG. Each box will need to be completed before moving onto the next (seems similar to S' `subclock`)

![image](https://user-images.githubusercontent.com/1354771/27090862-6955fbbc-502c-11e7-98ff-42cf01c79335.png)

Then, within each component, we effectively unroll the cycle into iterations. Each iteration runs in its own sub-subclock, so the calculation for `B@1` won't run until there are no more computations at time 0.

![image](https://user-images.githubusercontent.com/1354771/27091007-d7b7fee8-502c-11e7-916d-24d6a3ac15d2.png)

Eventually the node `C` won't emit a new value (effectively the base case in the recursion), which will allow that component to be complete, and then move onto `E`

Do you think this kind of approach is:

* Possible to build with S today?
* Possible to enhance S to support? (I'd be more than happy to contribute)
* Out of scope

Thanks again for your help

**Comment by adamhaile** [comment]
Hi @fenduru -
Sorry for the slow reply, as I'm off this week on vacation.  I'll be back next week.

I'm pretty sure this is accomplishable with current S, though I may not fully understand what you're trying to do.  In general, yes, being glitchless is a core feature of S, and subclocks are designed for this kind of "run to completion" scenario.  So I understand, in your diagrams, at t 0, B only depends on A.  At t > 0, B only depends on C?  Is that right?
-- Adam

**Comment by fenduru** [comment]
@adamhaile No worries! My diagram isn't totally accurate - I kind of conflated "dependencies" with "new value emitted". At t>0, B still depends on A, however we know that A never changes past that point so the only node that can trigger an update to B at t>0 is C.

Your original response gave me a lot of ideas/inspiration, so right now

**Comment by irisjay** [comment]
@fenduru Any news on how your approach is going? Interested to hear more

**Comment by adamhaile** [comment]
S does topological execution.  To be more precise, it doesn't guarantee that computations *start* in topological order, because it can't -- S's dependency graph is dynamic and changes each execution, due to conditional branches and higher-ordered signals (signals carrying signals).  Since we don't know what the graph will be prior to execution, we can't sort by it.  

What we can guarantee is that computations *complete* in topological order.  That means that when S detects that a computation is running before one of its sources has updated, it may pause the execution of the original computation to update the source before returning the source's new value and resuming execution of the original computation.

The algorithm behind this is pretty simple: S is a "two-pass" reactive implementation, meaning it does a first pass to mark all computations downstream of a modified data signal as invalid (i.e., it's not "height"-based, like it sounds like yours is).  The second pass it updates the computations.  The order of these updates actually isn't that critical, because if we start to read a computation that is still invalid, S executes and re-validates it then and there.  When S later encounters that computation at its originally scheduled time, it notes that it's already been revalidated and doesn't re-revalidate it.

Short of all this is that I believe S is already doing what you call the "ideal" scenario.

Here's some code that, if I understand you correctly, implements your topology:

```javascript
var input = S.data(0),
    lastE = S.data(0),
    A = S(() => input() + lastE()),
    B = S(() => A() + 1),
    C = S(() => B() + 1),
    D = S(() => C() + 1),
    E = S(() => D() + A() + 1),
    driver = S(() => E() > 20 || lastE(E()));
```

In this case, the looping continues until E() is greater than 20.

If you want to see how it executes, add a logging statement to record every time A() updates:

```javascript
S(() => console.log(A());
S.freeze(() => { input(0); lastE(0); });
```

**Comment by fenduru** [comment]
@adamhaile yes, in the sample you provided it will execute in the desired way, however by manually piping E back into `lastE` you're essentially encoding the solution to `MFAS` problem I described above. In other words you have manually broken the cycle into a DAG that will be run iteratively, rather than the system doing it for you.

The problem is this doesn't generalize though. Including additional inputs or feedback paths makes it far less obvious how to break the cycle down - it may even differ at runtime depending on which inputs actually emit a new signal. My particular use case complicates it even further as the exact shape of my graph isn't known until runtime.

I think if you are building a particular graph, and you know exactly how it should execute, then this approach may work well - though in that scenario it also makes a reactivity system less interesting as you could just as easily write the computation imperatively.
